msticpy/data/drivers/local_osquery_driver.py

```diff
@@ -0,0 +1,231 @@
+                    if "columns_atime" in test_df3.columns:
+                        test_df3["columns_atime"] = pd.to_datetime(
+                            test_df3["columns_atime"], unit="s", origin="unix"
+                        )
+                    if "columns_ctime" in test_df3.columns:
+                        test_df3["columns_ctime"] = pd.to_datetime(
+                            test_df3["columns_ctime"], unit="s", origin="unix"
+                        )
+                    if "columns_mtime" in test_df3.columns:
+                        test_df3["columns_mtime"] = pd.to_datetime(
+                            test_df3["columns_mtime"], unit="s", origin="unix"
+                        )
+
+                    if not isinstance(test_df3, pd.DataFrame):
+                        continue
+                    df_schema = test_df3.dtypes
+                    new_schema[df_fname2] = {
+                        key: dtype.name for key, dtype in df_schema.to_dict().items()
+                    }
+                self._schema = new_schema | schema_backup
+            else:
+                if not isinstance(test_df, pd.DataFrame):
+                    continue
+                df_schema = test_df.dtypes
+                self._schema[df_fname] = {
+                    key: dtype.name for key, dtype in df_schema.to_dict().items()
+                }
+
+        return self._schema
+
+    def query(
+        self, query: str, query_source: QuerySource = None, **kwargs
+    ) -> Union[pd.DataFrame, Any]:
+        """
+        Execute query string and return DataFrame of results.
+
+        Parameters
+        ----------
+        query : str
+            The query to execute
+        query_source : QuerySource
+            The query definition object
+
+        Returns
+        -------
+        Union[pd.DataFrame, results.ResultSet]
+            A DataFrame (if successfull) or
+            the underlying provider result if an error.
+
+        """
+        del kwargs
+        query_name = query_source.name if query_source else query
+        file_path = self.data_files.get(query.casefold())
+        # print(f"Data file path {file_path} (data_files: {self.data_files}).")
+        if not file_path:
+            raise FileNotFoundError(
+                f"Data file ({query}) for query {query_name} not found."
+            )
+        if file_path.endswith("log") or file_path.endswith("json"):
+            # We can't use directly read_json as we need to extract nested json
+            # Split log files in multiple queries table here
+            # Likely resource intensive and better way to do.
+            # Likely issue, multiple log files can contain same query mostly
+            # because of log rotation
+            try:
+                with open(file_path, mode="r", encoding="utf-8") as logfile:
+                    json_out = {"lines": []}
+                    while True:
+                        next_line = logfile.readline()
```

Comment:
Would need to test/time out.

I think it would likely be more efficient to cache data after initial read. not familiar with pandas caching.
From quick search,
* native read_pickle/to_pickle. I believe you already use this, so likely the choice.
* native pandas.HDFStore()
* https://github.com/statnett/data_cache as decorator but does not seem maintained
* https://github.com/susmitpy/cache_df, also maintenance issue

Reply:
I've implemented something like this in the driver - with a `cache_file` parameter

Reply:
Thanks a lot!